compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -1123,7 +1123,7 @@ pub(super) fn transmute_immediate<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     // While optimizations will remove no-op transmutes, they might still be
     // there in debug or things that aren't no-op in MIR because they change
     // the Rust type but not the underlying layout/niche.
-    if from_scalar == to_scalar && from_backend_ty == to_backend_ty {
+    if from_scalar == to_scalar {
         return imm;
     }
```

Comment:
I believe that this change is incorrect. We can't just implicitly convert between types that have the same scalar representation, but a different backend type. For instance

https://godbolt.org/z/PaaGWTv5a

```rust
#![feature(repr_simd)]

#[repr(simd)]
#[derive(Clone, Copy)]
struct S([i64; 1]);

#[no_mangle]
pub extern "C" fn single_element_simd(b: S) -> i64 {
    unsafe { std::mem::transmute(b) }
}
```

generates broken IR: 

```llvm
define noundef i64 @single_element_simd(<1 x i64> %b) unnamed_addr {
start:
  ret <1 x i64> %b
}
```

Reply:
It's very strange that a function that is supposed to work on all "immediates" just takes `abi::Scalar`. That doesn't fully describe an immediate. Either the function should be renamed to `transmute_scalar` (and never be called for non-scalar types), or it should take the full `BackendRepr` so that it has a clue what the two immediates are.